### PR TITLE
Dynamic Dashboard: Coupons card

### LINKED ISSUE #12716: Top 3 active coupons card UI
image

- [x] First iteration of Most Active coupons card UI work. #12733
- [ ] Empty view when there are no active coupons. 
- [x] Coupon detail popup has a "More" (Ellipsis) menu. Test that it works as expected.  #12733

### LINKED ISSUE #12728: Top 3 active coupons card logic
- [ ] Fetch and show most active coupons from remote 
- [ ] Unit tests for `MostActiveCouponsCardViewModel`

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #12716 
Closes: #12728
Part of: #12698

## Description

Completes the UI and logic for the coupons card in dashboard.

Changes
- Action and remote method for loading coupons using coupon IDs.
- Fetch coupon report and coupons from remote and display those.
- Empty view and content unavailable view.
- Unit tests.

## Testing instructions

- Enable the feature flag `dynamicDashboardM2` and build the app.
- Enable the Coupon card on the dashboard.
- Confirm that the card now displays the 3 most active coupons for the selected time range. 
- Validate that the UI matches the design.
- Change the time range and validate the coupons are listed based on the selected time range.
- If there are no coupons available for a selected time range an empty view with the message "No coupon usage during this period" should be displayed.
- Tap on any coupon and validate that the coupon details view is displayed.
- Try editing or deleting the coupon and ensure that it works as expected. 
- Navigate back to the dashboard. Tap View all coupons and confirm that the coupons list view is displayed.
- Disconnect the Internet connect and reload the card. Confirm that the error state is displayed.
- Disable "WooCommerce Analytics" toggle from your wp-admin page. Reload the card and ensure that error view is displayed asking you to check WooCommerce version/settings or contact support.
- Tap the ellipsis button and dismiss the card. Confirm that event `dynamic_dashboard_hide_card_tapped` is tracked with type `coupons`.

## Screenshots

Error cases
| WooCommerce Analytics disabled | No active coupons available for the selected time range | No internet  |
|--------|--------|--------|
| image | image | image | 

wp-admin page "Enable WooCommerce Analytics" toggle

Screenshot 2024-05-16 at 3 16 36 PM

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift [hunk 1]
```diff
@@ -110,6 +124,17 @@ private extension MostActiveCouponsCard {
...
     }
 
+    var emptyView: some View {
+        HStack {
+            Spacer()
+
+            MostActiveCouponsEmptyView()
+
+            Spacer()
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift]: Just a suggestion: you can avoid adding `Spacer()` by setting the width of the view to the maximum - that should also push it to the center.

```suggestion
        MostActiveCouponsEmptyView()
            .frame(maxWidth: .infinity)
```

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift [hunk 2]
```diff
@@ -159,6 +184,37 @@ private extension MostActiveCouponsCard {
...
+            Button(Localization.ContentUnavailable.buttonTitle) {
+                showingSupportForm = true
+            }
+            .buttonStyle(SecondaryButtonStyle())
+        }
+        .frame(maxWidth: .infinity)
+    }
+
+    var supportForm: some View {
+        NavigationView {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift]: Please refrain from using `NavigationView` because it's deprecating soon. `NavigationStack` should be a better alternative.

**Review by itsmeichigo — APPROVED** [review]
This works great! I noticed a few minor points:

- Should we reorder the items by usage count? In my case, the item with the lower usage count is displayed first, which doesn't make sense when we want to emphasize active items more:



Another minor issue: we should keep the divider between the empty state and the view all button too keep them separate IMO:



Lastly, I wonder if we still fetch items from storage so that we can display the results earlier. The coupon loading request can still be triggered in the background to keep the stored items updated, I just thought if we could display cached items right away that would enhance the UX.

I'm pre-approving the PR so please feel free to merge after addressing the above issues.

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/5079a5ddf9ad1c8fd39245159d29119a55a6b1d7

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/b46ee06380fc126cd18f95b891ce15d8d346b339

**Comment by selanthiraiyan** [comment]
Thanks for the review and suggestions, @itsmeichigo! 

---

> Should we reorder the items by usage count? In my case, the item with the lower usage count is displayed first, which doesn't make sense when we want to emphasize active items more:

Good catch. The usage count we were showing was the total usage count. I changed that and started displaying the usage count for the selected time range in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/c17f296fa7a74d97c8a0d83a37ea55079b2ba994. 

> Another minor issue: we should keep the divider between the empty state and the view all button too keep them separate IMO:

Fixed in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/5079a5ddf9ad1c8fd39245159d29119a55a6b1d7

---

> Lastly, I wonder if we still fetch items from storage so that we can display the results earlier. The coupon loading request can still be triggered in the background to keep the stored items updated, I just thought if we could display cached items right away that would enhance the UX.

Great suggestion. I changed the loading behaviour and started reading coupons from storage in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/409ef95feccd25fd5d1d946c5f652dbbd2f47ae4.

Please take another look at the PR when you can. 🙇

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift [hunk 3]
```diff
@@ -159,39 +182,36 @@ private extension MostActiveCouponsCard {
...
+                        showingSupportForm = false
                     }
                 }
-                .padding(.horizontal, MostActiveCouponsCard.Layout.padding)
-
-                Divider()
-                    .padding(.leading, MostActiveCouponsCard.Layout.padding)
             }
         }
+        .navigationViewStyle(.stack)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift]: Nit: this is redundant for `NavigationStack`

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift [hunk 4]
```diff
@@ -66,9 +67,27 @@ final class MostActiveCouponsCardViewModel: ObservableObject {
...
+                /// If all or some coupons are not available in storage
+                /// then fetch from remote
+                if rows.count != couponReports.count {
+                    try await synchronizeCoupons(for: couponReports)
+                } else {
+                    /// Refresh coupons in background
+                    Task { @MainActor in
+                        try await synchronizeCoupons(for: couponReports)
+                    }
+                }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift:89]: I found a possible edge case with my test store. The coupon reports return 3 items and one of them has already been removed so the sync call cannot find that one coupon => the code always fall to the `else` case. I believe this is not a serious case, but that makes me wonder how often this happens and if it makes our storage usage become a waste. I'll leave this for you to decide if we should go ahead without storage (sorry for the back and forth 🙇).

**Review by itsmeichigo — APPROVED** [review]
This still LGTM, but I found an edge case in a comment below.

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift (hunk 4 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCardViewModel.swift:89]: Nice find, Huong! I will keep the storage-related logic as it doesn't create any issues. (Even though it won't be useful for stores that have deleted any of the most active coupons.)

File: WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Dashboard/Coupons/MostActiveCouponsCard.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12755/commits/d8bbf57833d1bbb15db8bacca89182b18b65f705
